port/raspberrypi/rp2xxx/src/hal/pio.zig

```diff
@@ -3,644 +3,92 @@ const std = @import("std");
+pub const StateMachineInitOptions = chip_specific.StateMachineInitOptions;
+pub const LoadAndStartProgramOptions = chip_specific.LoadAndStartProgramOptions;
+pub const ClkDivOptions = common.ClkDivOptions;
 
 const gpio = @import("gpio.zig");
 const resets = @import("resets.zig");
 const hw = @import("hw.zig");
 pub const assembler = @import("pio/assembler.zig");
 const encoder = @import("pio/assembler/encoder.zig");
-
-// global state for keeping track of used things
-var used_instruction_space: [2]u32 = [_]u32{ 0, 0 };
-var claimed_state_machines: [2]u4 = [_]u4{ 0, 0 };
-
-pub const Instruction = encoder.Instruction;
-pub const Program = assembler.Program;
 pub const assemble = assembler.assemble;
 
-pub const Fifo = enum {
-    tx,
-    rx,
-};
-
-pub const StateMachine = enum(u2) {
-    sm0,
-    sm1,
-    sm2,
-    sm3,
-
-    pub const Regs = extern struct {
-        clkdiv: @TypeOf(PIO0.SM0_CLKDIV),
-        execctrl: @TypeOf(PIO0.SM0_EXECCTRL),
-        shiftctrl: @TypeOf(PIO0.SM0_SHIFTCTRL),
-        addr: @TypeOf(PIO0.SM0_ADDR),
-        instr: @TypeOf(PIO0.SM0_INSTR),
-        pinctrl: @TypeOf(PIO0.SM0_PINCTRL),
-    };
-
-    comptime {
-        assert(@sizeOf([2]Regs) == (4 * 6 * 2));
-    }
-};
-
-pub const Irq = enum {
-    irq0,
-    irq1,
-
-    pub const Regs = extern struct {
-        enable: @TypeOf(PIO0.IRQ0_INTE),
-        force: @TypeOf(PIO0.IRQ0_INTF),
-        status: @TypeOf(PIO0.IRQ0_INTS),
-    };
-
-    comptime {
-        assert(@sizeOf([2]Regs) == (3 * 4 * 2));
-    }
-
-    pub const Source = enum {
-        rx_not_empty,
-        tx_not_full,
-        // TODO: determine what this does, is it just a combination of the
-        // first two, or is it other things?
-        statemachine,
-    };
-};
-
-pub const ClkDivOptions = struct {
-    int: u16 = 1,
-    frac: u8 = 0,
-
-    pub fn from_float(div: f32) ClkDivOptions {
-        const fixed = @as(u24, @intFromFloat(div * 256));
-        return ClkDivOptions{
-            .int = @as(u16, @truncate(fixed >> 8)),
-            .frac = @as(u8, @truncate(fixed)),
-        };
-    }
-};
-
-pub const ExecOptions = struct {
-    jmp_pin: u5 = 0,
-    wrap: u5 = 31,
-    wrap_target: u5 = 0,
-    side_pindir: bool = false,
-    side_set_optional: bool = false,
-};
-
-pub const ShiftOptions = struct {
-    autopush: bool = false,
-    autopull: bool = false,
-    in_shiftdir: Direction = .right,
-    out_shiftdir: Direction = .right,
-    /// 0 means full 32-bits
-    push_threshold: u5 = 0,
-    /// 0 means full 32-bits
-    pull_threshold: u5 = 0,
-    join_tx: bool = false,
-    join_rx: bool = false,
-
-    pub const Direction = enum(u1) {
-        left,
-        right,
-    };
-};
-
-pub fn PinMapping(comptime Count: type) type {
-    return struct {
-        base: u5 = 0,
-        count: Count = 0,
-    };
+pub const ShiftOptions = chip_specific.ShiftOptions;
+// TODO: THis works but we probably should use the chip_specific stuff
```

Comment:
Not sure if you'd rather put _everything_ unique in its own file, or use this stuff when able.